File: concurrency-named-lock/src/main/java/io/github/hyeonic/concurrencynamedlock/domain/Product.java
package io.github.hyeonic.concurrencynamedlock.domain;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.Version;

@Entity
public class Product {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "name")
    private String name;

    @Column(name = "quantity")
    private Long quantity;

    @Version
    private Long version;

    protected Product() {
    }

    public Product(final String name, final Long quantity) {
        this.name = name;
        this.quantity = quantity;
    }

    public void decrease(final Long quantity) {
        if (this.quantity - quantity < 0) {
            throw new RuntimeException("수량은 0개 미만이 될 수 없습니다.");
        }

        this.quantity -= quantity;
    }

    public Long getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public Long getQuantity() {
        return quantity;
    }
}

File: redis-pub-sub/src/main/kotlin/c/hyeoni/redispubsub/domain/product/Product.kt
package c.hyeoni.redispubsub.domain.product

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.Version
import org.springframework.context.ApplicationEvent
import org.springframework.data.domain.AbstractAggregateRoot

@Entity
class Product(
    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long = 0,

    @Column(unique = true, nullable = false)
    var name: String,

    @Column(unique = true, nullable = false)
    var description: String
) : AbstractAggregateRoot<Product>() {

    @Version
    var versionNo: Long = 0L

    fun notifyChanged() {
        if (id != 0L) {
            this.registerEvent(ProductChangedEvent(this.id))
        }
    }
}

data class ProductChangedEvent(val id: Long): ApplicationEvent(id)
